session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java

```diff
@@ -0,0 +1,670 @@
+              this.wait(1000);
+              if (System.currentTimeMillis() - start > 60_000) {
+                logger.warn(
+                    "the SessionPool has wait for {} seconds to get a new connection: {}:{} with {}, {}",
+                    (System.currentTimeMillis() - start) / 1000, ip, port, user, password);
+              }
+            } catch (InterruptedException e) {
+              logger.error("the SessionPool is damaged", e);
+              Thread.currentThread().interrupt();
+            }
+            session = queue.poll();
+          }
+        }
+        if (session != null) {
+          return session;
+        }
+      }
+      if (logger.isDebugEnabled()) {
+        logger.error("Create a new Session {}, {}, {}, {}", ip, port, user, password);
+      }
+      session = new Session(ip, port, user, password, fetchSize);
+      session.open();
+      return session;
+    }
+  }
+
+  public int currentAvailableSize() {
+    return queue.size();
+  }
+
+  public int currentOccupiedSize() {
+    return occupied.size();
+  }
+
+  private void putBack(Session session) {
+    queue.push(session);
+    synchronized (this) {
+      this.notifyAll();
+    }
+  }
+
+  private void occupy(Session session) {
+    occupied.put(session, session);
+  }
+
+  /**
+   * close all connections in the pool
+   */
+  public synchronized void close() {
+    for (Session session : queue) {
+      try {
+        session.close();
+      } catch (IoTDBSessionException e) {
+        //do nothing
+      }
+    }
+    for (Session session : occupied.keySet()) {
+      try {
+        session.close();
+      } catch (IoTDBSessionException e) {
+        //do nothing
+      }
+    }
+    queue.clear();
+    occupied.clear();
+  }
+
+  public void closeResultSet(SessionDataSetWrapper wrapper) throws SQLException {
+    boolean putback = true;
+    try {
+      wrapper.sessionDataSet.closeOperationHandle();
+    } catch (SQLException e) {
+      if (e.getCause() instanceof TException) {
+        // the connection is broken.
+        removeSession();
+        putback = false;
+      } else {
+        throw e;
+      }
+    } finally {
+      Session session = occupied.remove(wrapper.session);
+      if (putback && session != null) {
+        putBack(wrapper.session);
+      }
+    }
+  }
+
+  private synchronized void removeSession() {
+    if (logger.isDebugEnabled()) {
+      logger.error("Remove a broken Session {}, {}, {}, {}", ip, port, user, password);
+    }
+    size--;
+  }
+
+  private void closeSession(Session session) {
+    if (session != null) {
+      try {
+        session.close();
+      } catch (Exception e2) {
+        //do nothing. We just want to guarantee the session is closed.
+      }
+    }
+  }
+
+  /**
+   * use batch interface to insert sorted data times in row batch must be sorted before!
+   *
+   * @param rowBatch data batch
+   */
+  public TSExecuteBatchStatementResp insertSortedBatch(RowBatch rowBatch)
```

Comment:
Is it needed to export the methods of Session in this Pool? the codes are double...

Reply:
>  the codes are double...
Ah?

Reply:
the interfaces are double...